node/network/availability-distribution/src/requester/mod.rs

```diff
@@ -238,7 +224,7 @@ impl Stream for Requester {
 async fn query_occupied_cores<Context>(
 	ctx: &mut Context,
 	relay_parent: Hash,
-) -> Result<Vec<OccupiedCore>, NonFatalError>
+) -> Result<Vec<OccupiedCore>, Error>
```

Comment:
why this change from `NonFatalError`?

Reply:
It was inconsistent and did not feel useful at all. (Like we still did not fail on lots of fatal `Error`s)  The distinction felt pretty arbitrary, also what I got, we basically never want to fail, which makes that distinction even less useful. If there are circumstances we really want to crash, we can panic, which should be pretty much the same as propagating `Err` to the overseer.

Reply:
I guess the distinction was about whether to log an error or not. We probably don't want to log a runtime api error for example.

Reply:
I think we should do a `debug_assert` on logic errors or if really bad an `assert`, this should get us the same thing, but with less complexity.

Reply:
I see 3 classes of errors:
* NonFatal - log and continue - atm it's only RuntimeApiError
* Logic - debug_assert/assert and warn
* ChannelClosed - a subsystem is down - we're shutting down - early exit

Reply:
panicking is usually bad because we may trigger double panic which would lead to a segfault which may be hard to debug and some subsystems might prefer to have a clean shutdown to e.g. flush a db transaction

Reply:
Hmm. Then I think we could do a trait `IsFatal`with a function `isFatal` and have a generic function that just logs non fatal errors and will propagate fatal ones, that would be the least invasive method I guess. 

Like, changing what is fatal and what is not can be done in a single place and we could have a default implementation for `isFatal` which just always returns `false`.

Reply:
There were some attempts to do this systematically (starting with #1888), but we never got around enforcing this on a compile time level as there were always more important stuff to do :P

I agree though that'd be really nice to have in one place.